Orange/widgets/utils/messagewidget.py

```diff
@@ -0,0 +1,561 @@
+import sys
+import enum
+import base64
+from itertools import chain
+from operator import attrgetter
+from xml.sax.saxutils import escape
+from collections import OrderedDict
+# pylint: disable=unused-import
+from typing import (
+    NamedTuple, Tuple, List, Dict, Iterable, Union, Optional, Hashable
+)
+
+from AnyQt.QtCore import Qt, QSize, QBuffer
+from AnyQt.QtGui import (
+    QIcon, QPixmap, QPainter, QPalette, QLinearGradient, QBrush, QPen
+)
+from AnyQt.QtWidgets import (
+    QWidget, QLabel, QSizePolicy, QStyle, QHBoxLayout, QMessageBox,
+    QMenu, QWidgetAction, QStyleOption, QApplication
+)
+from AnyQt.QtCore import pyqtSignal as Signal
+
+__all__ = ["Message", "MessagesWidget"]
+
+
+def image_data(pm):
+    # type: (QPixmap) -> str
+    """
+    Render the contents of the pixmap as a data URL (RFC-2397)
+
+    Parameters
+    ----------
+    pm : QPixmap
+
+    Returns
+    -------
+    datauri : str
+    """
+    pm = QPixmap(pm)
+    device = QBuffer()
+    assert device.open(QBuffer.ReadWrite)
+    pm.save(device, b'png')
+    device.close()
+    data = bytes(device.data())
+    payload = base64.b64encode(data).decode("ascii")
+    return "data:image/png;base64," + payload
```

Comment:
This could be also useful elsewhere. Do we have a suitable place for such functions? (If not, ignore.)